File: sdk-go/processes.go
package boxkite

import (
	"bufio"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strconv"
	"strings"
)

// StartProcessOptions carries the optional parameters for StartProcess.
type StartProcessOptions struct {
	Description *string
	// MaxRuntimeSeconds is a hard ceiling on how long the process may run
	// before being force-killed. Defaults to 3600.
	MaxRuntimeSeconds int
	// ExposePort, if set, makes this process's listening port reachable
	// via a preview URL (see CreatePreviewURL).
	ExposePort *int
}

// ProcessStartResult is the response shape from StartProcess.
type ProcessStartResult struct {
	ProcessID string `json:"process_id"`
	Status    string `json:"status"`
	StartedAt string `json:"started_at"`
}

// StartProcess starts a background process that keeps running after this
// call returns (POST /v1/sandboxes/{id}/processes). Distinct from Exec,
// which is one-shot request/response bounded by its own timeout: poll the
// returned ProcessID's output with GetProcessOutput, feed it input with
// SendProcessInput, and stop it with StopProcess.
func (c *Client) StartProcess(ctx context.Context, sessionID, command string, opts *StartProcessOptions) (*ProcessStartResult, error) {
	maxRuntimeSeconds := 3600
	body := map[string]any{"command": command}
	if opts != nil {
		if opts.MaxRuntimeSeconds != 0 {
			maxRuntimeSeconds = opts.MaxRuntimeSeconds
		}
		if opts.Description != nil {
			body["description"] = *opts.Description
		}
		if opts.ExposePort != nil {
			body["expose_port"] = *opts.ExposePort
		}
	}
	body["max_runtime_seconds"] = maxRuntimeSeconds
	var out ProcessStartResult
	path := fmt.Sprintf("/v1/sandboxes/%s/processes", url.PathEscape(sessionID))
	if err := c.doJSON(ctx, "POST", path, body, &out, nil); err != nil {
		return nil, err
	}
	return &out, nil
}

// ProcessInfo is one tracked background process, as returned by
// ListProcesses.
type ProcessInfo struct {
	ProcessID   string  `json:"process_id"`
	Command     string  `json:"command"`
	Description *string `json:"description"`
	Status      string  `json:"status"`
	StartedAt   string  `json:"started_at"`
	ExitCode    *int    `json:"exit_code"`
}

// ProcessListResult is the response shape from ListProcesses.
type ProcessListResult struct {
	Processes []ProcessInfo `json:"processes"`
}

// ListProcesses returns every background process currently tracked for
// this session (GET /v1/sandboxes/{id}/processes).
func (c *Client) ListProcesses(ctx context.Context, sessionID string) (*ProcessListResult, error) {
	var out ProcessListResult
	path := fmt.Sprintf("/v1/sandboxes/%s/processes", url.PathEscape(sessionID))
	if err := c.doJSON(ctx, "GET", path, nil, &out, nil); err != nil {
		return nil, err
	}
	return &out, nil
}

// ProcessOutputResult is the response shape from GetProcessOutput.
type ProcessOutputResult struct {
	Status      string `json:"status"`
	StdoutChunk string `json:"stdout_chunk"`
	NextOffset  int64  `json:"next_offset"`
	Truncated   bool   `json:"truncated"`
	ExitCode    *int   `json:"exit_code"`
}

// GetProcessOutput polls a background process's output since a given byte
// offset (GET /v1/sandboxes/{id}/processes/{processId}/output).
// Polling-style, not streaming. sinceOffset (from a previous call's
// NextOffset, or 0 the first time) lets you fetch only the new output
// since your last check.
func (c *Client) GetProcessOutput(ctx context.Context, sessionID, processID string, sinceOffset int64) (*ProcessOutputResult, error) {
	q := newQuery()
	q.Set("since_offset", strconv.FormatInt(sinceOffset, 10))
	opts := &requestOptions{query: q}
	var out ProcessOutputResult
	path := fmt.Sprintf("/v1/sandboxes/%s/processes/%s/output", url.PathEscape(sessionID), url.PathEscape(processID))
	if err := c.doJSON(ctx, "GET", path, nil, &out, opts); err != nil {
		return nil, err
	}
	return &out, nil
}

// ProcessStreamEvent is one event from StreamProcessOutput. Type is "output"
// (StdoutChunk/NextOffset/Truncated populated) for each new chunk, then "exit"
// (Status/ExitCode populated) once the process finishes.
type ProcessStreamEvent struct {
	Type        string `json:"type"`
	StdoutChunk string `json:"stdout_chunk"`
	NextOffset  int64  `json:"next_offset"`
	Truncated   bool   `json:"truncated"`
	Status      string `json:"status"`
	ExitCode    *int   `json:"exit_code"`
}

// ProcessStream streams a background process's stdout as Server-Sent Events
// (GET /v1/sandboxes/{id}/processes/{processId}/stream), the streaming
// counterpart to GetProcessOutput. Use it like bufio.Scanner:
//
//	stream, err := client.StreamProcessOutput(ctx, sessionID, processID, 0)
//	if err != nil { ... }
//	defer stream.Close()
//	for stream.Next() {
//		ev := stream.Event()
//		// ev.Type == "output" | "exit"
//	}
//	if err := stream.Err(); err != nil { ... }
type ProcessStream struct {
	resp    *http.Response
	scanner *bufio.Scanner
	current ProcessStreamEvent
	err     error
}

// StreamProcessOutput opens the SSE output stream for a background process.
// The caller must call Close on the returned *ProcessStream once done.
func (c *Client) StreamProcessOutput(ctx context.Context, sessionID, processID string, sinceOffset int64) (*ProcessStream, error) {
	path := fmt.Sprintf("/v1/sandboxes/%s/processes/%s/stream", url.PathEscape(sessionID), url.PathEscape(processID))
	reqURL := c.baseURL + path + "?since_offset=" + strconv.FormatInt(sinceOffset, 10)
	req, err := http.NewRequestWithContext(ctx, "GET", reqURL, nil)
	if err != nil {
		return nil, &ConnectionError{Message: err.Error(), Err: err}
	}
	req.Header.Set("Authorization", "Bearer "+c.apiKey)

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, &ConnectionError{Message: err.Error(), Err: err}
	}
	if resp.StatusCode >= 400 {
		body, _ := io.ReadAll(resp.Body)
		resp.Body.Close()
		return nil, apiErrorFromResponse(resp.StatusCode, body)
	}

	scanner := bufio.NewScanner(resp.Body)
	scanner.Buffer(make([]byte, 0, 64*1024), 8*1024*1024)
	return &ProcessStream{resp: resp, scanner: scanner}, nil
}

// Next advances to the next streamed event, blocking until one arrives. It
// returns false when the stream ends -- check Err to distinguish a clean end
// from an error.
func (s *ProcessStream) Next() bool {
	var dataLines []string
	for s.scanner.Scan() {
		line := s.scanner.Text()
		if line == "" {
			if len(dataLines) == 0 {
				continue
			}
			joined := strings.Join(dataLines, "\n")
			var ev ProcessStreamEvent
			if err := json.Unmarshal([]byte(joined), &ev); err != nil {
				s.err = fmt.Errorf("boxkite: decoding SSE event: %w", err)
				return false
			}
			s.current = ev
			return true
		}
		if strings.HasPrefix(line, "data:") {
			dataLines = append(dataLines, strings.TrimPrefix(strings.TrimPrefix(line, "data:"), " "))
		}
	}
	if err := s.scanner.Err(); err != nil {
		s.err = err
	}
	return false
}

// Event returns the most recent event produced by Next.
func (s *ProcessStream) Event() ProcessStreamEvent { return s.current }

// Err returns the first error encountered while streaming, if any.
func (s *ProcessStream) Err() error { return s.err }

// Close closes the underlying HTTP response body, ending the stream.
func (s *ProcessStream) Close() error { return s.resp.Body.Close() }

// ProcessInputResult is the response shape from SendProcessInput.
type ProcessInputResult struct {
	BytesWritten int `json:"bytes_written"`
}

// SendProcessInput writes to a tracked background process's stdin pipe
// (POST /v1/sandboxes/{id}/processes/{processId}/input).
func (c *Client) SendProcessInput(ctx context.Context, sessionID, processID, data string) (*ProcessInputResult, error) {
	body := map[string]any{"data": data}
	var out ProcessInputResult
	path := fmt.Sprintf("/v1/sandboxes/%s/processes/%s/input", url.PathEscape(sessionID), url.PathEscape(processID))
	if err := c.doJSON(ctx, "POST", path, body, &out, nil); err != nil {
		return nil, err
	}
	return &out, nil
}

// ProcessStopResult is the response shape from StopProcess.
type ProcessStopResult struct {
	Status   string `json:"status"`
	ExitCode *int   `json:"exit_code"`
}

// StopProcess stops a tracked background process (SIGTERM, then SIGKILL
// if it doesn't exit within a short grace period)
// (POST /v1/sandboxes/{id}/processes/{processId}/stop).
func (c *Client) StopProcess(ctx context.Context, sessionID, processID string) (*ProcessStopResult, error) {
	var out ProcessStopResult
	path := fmt.Sprintf("/v1/sandboxes/%s/processes/%s/stop", url.PathEscape(sessionID), url.PathEscape(processID))
	if err := c.doJSON(ctx, "POST", path, nil, &out, nil); err != nil {
		return nil, err
	}
	return &out, nil
}

File: sdk-rust/src/processes.rs
//! Background processes/sessions (`docs/PROCESS-SESSIONS-DESIGN.md`):
//! `POST /v1/sandboxes/{id}/processes` and its `/{process_id}/output`,
//! `/{process_id}/input`, `/{process_id}/stop` siblings, plus
//! `GET /v1/sandboxes/{id}/processes`. Mirrors `sdk-python`'s
//! `start_process`/`list_processes`/`get_process_output`/
//! `send_process_input`/`stop_process`.
//!
//! Distinct from [`crate::Client::exec`]: `exec` is one-shot request/
//! response bounded by its own timeout; these routes track a process across
//! multiple calls. Polling-style, not streaming -- the control plane's own
//! route comments mark SSE/streaming process output as a separate, later
//! phase, not part of this route set.

use std::pin::Pin;

use futures_util::{Stream, StreamExt};
use reqwest::Method;
use reqwest_eventsource::{Event, EventSource};
use serde::{Deserialize, Serialize};

use crate::client::Client;
use crate::error::BoxkiteError;

/// `max_runtime_seconds` is required by the server
/// (`SandboxProcessStartRequest`) with a 4h ceiling
/// (`SANDBOX_PROCESS_MAX_RUNTIME_SECONDS_CEILING`) -- this is only this
/// crate's client-side default when the caller doesn't set one, matching
/// `sdk-python`/`sdk-js`/`sdk-go`'s own default of 3600. The ceiling itself
/// is enforced server-side and deliberately not duplicated here.
const DEFAULT_MAX_RUNTIME_SECONDS: i64 = 3600;

/// Builder for `POST /v1/sandboxes/{id}/processes`'s optional fields.
#[derive(Debug, Clone)]
pub struct StartProcessOptions {
    description: Option<String>,
    max_runtime_seconds: i64,
    expose_port: Option<i32>,
}

impl Default for StartProcessOptions {
    fn default() -> Self {
        Self {
            description: None,
            max_runtime_seconds: DEFAULT_MAX_RUNTIME_SECONDS,
            expose_port: None,
        }
    }
}

impl StartProcessOptions {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn description(mut self, description: impl Into<String>) -> Self {
        self.description = Some(description.into());
        self
    }

    /// Hard ceiling on how long the process may run before being
    /// force-killed, in seconds. Defaults to 3600; capped server-side at 4h.
    pub fn max_runtime_seconds(mut self, max_runtime_seconds: i64) -> Self {
        self.max_runtime_seconds = max_runtime_seconds;
        self
    }

    /// Opt-in port to expose via a signed preview URL once the process is
    /// listening -- see `docs/NETWORK-INGRESS-DESIGN.md` and
    /// `POST /v1/sandboxes/{id}/preview/{port}`. Leave unset for a normal,
    /// fully network-isolated background process.
    pub fn expose_port(mut self, expose_port: i32) -> Self {
        self.expose_port = Some(expose_port);
        self
    }
}

#[derive(Serialize)]
struct StartProcessBody<'a> {
    command: &'a str,
    #[serde(skip_serializing_if = "Option::is_none")]
    description: Option<&'a str>,
    max_runtime_seconds: i64,
    #[serde(skip_serializing_if = "Option::is_none")]
    expose_port: Option<i32>,
}

/// `POST /v1/sandboxes/{id}/processes`'s response.
#[derive(Debug, Clone, Deserialize)]
pub struct ProcessStartResult {
    pub process_id: String,
    pub status: String,
    pub started_at: String,
}

/// One tracked background process (`SandboxProcessInfo`), as returned by
/// [`Client::list_processes`].
#[derive(Debug, Clone, Deserialize)]
pub struct ProcessInfo {
    pub process_id: String,
    pub command: String,
    pub description: Option<String>,
    pub status: String,
    pub started_at: String,
    pub exit_code: Option<i32>,
    pub expose_port: Option<i32>,
}

/// `GET /v1/sandboxes/{id}/processes`'s response.
#[derive(Debug, Clone, Deserialize)]
pub struct ProcessListResult {
    pub processes: Vec<ProcessInfo>,
}

/// `GET /v1/sandboxes/{id}/processes/{process_id}/output`'s response.
#[derive(Debug, Clone, Deserialize)]
pub struct ProcessOutputResult {
    pub status: String,
    pub stdout_chunk: String,
    pub next_offset: i64,
    pub truncated: bool,
    pub exit_code: Option<i32>,
}

/// One event from `GET /v1/sandboxes/{id}/processes/{process_id}/stream`: an
/// [`Output`](ProcessStreamEvent::Output) per new stdout chunk, then a
/// terminal [`Exit`](ProcessStreamEvent::Exit) once the process finishes.
#[derive(Debug, Clone, Deserialize)]
#[serde(tag = "type", rename_all = "lowercase")]
pub enum ProcessStreamEvent {
    Output {
        stdout_chunk: String,
        next_offset: i64,
        truncated: bool,
    },
    Exit {
        status: String,
        exit_code: Option<i32>,
    },
}

/// `POST /v1/sandboxes/{id}/processes/{process_id}/input`'s response.
#[derive(Debug, Clone, Deserialize)]
pub struct ProcessInputResult {
    pub bytes_written: i64,
}

/// `POST /v1/sandboxes/{id}/processes/{process_id}/stop`'s response.
#[derive(Debug, Clone, Deserialize)]
pub struct ProcessStopResult {
    pub status: String,
    pub exit_code: Option<i32>,
}

impl Client {
    /// `POST /v1/sandboxes/{session_id}/processes` -- start a background
    /// process (a dev server, a test watcher, a long build, a REPL) that
    /// keeps running after this call returns. Not reachable over the
    /// network from any other call by default -- same per-exec network
    /// isolation applies here -- unless `options.expose_port(..)` is set.
    pub async fn start_process(
        &self,
        session_id: &str,
        command: &str,
        options: StartProcessOptions,
    ) -> Result<ProcessStartResult, BoxkiteError> {
        let body = StartProcessBody {
            command,
            description: options.description.as_deref(),
            max_runtime_seconds: options.max_runtime_seconds,
            expose_port: options.expose_port,
        };
        let builder = self
            .request(
                Method::POST,
                &format!("/v1/sandboxes/{session_id}/processes"),
            )
            .json(&body);
        self.send(builder).await
    }

    /// `GET /v1/sandboxes/{session_id}/processes` -- every background
    /// process currently tracked for this session (running, exited, or
    /// stopped).
    pub async fn list_processes(
        &self,
        session_id: &str,
    ) -> Result<ProcessListResult, BoxkiteError> {
        let builder = self.request(
            Method::GET,
            &format!("/v1/sandboxes/{session_id}/processes"),
        );
        self.send(builder).await
    }

    /// `GET /v1/sandboxes/{session_id}/processes/{process_id}/output` --
    /// poll a background process's output since a given byte offset.
    /// Polling-style, not streaming. `since_offset` (from a previous call's
    /// `next_offset`, or 0 the first time) fetches only the new output
    /// since the last check.
    pub async fn get_process_output(
        &self,
        session_id: &str,
        process_id: &str,
        since_offset: i64,
    ) -> Result<ProcessOutputResult, BoxkiteError> {
        let builder = self
            .request(
                Method::GET,
                &format!("/v1/sandboxes/{session_id}/processes/{process_id}/output"),
            )
            .query(&[("since_offset", since_offset.to_string())]);
        self.send(builder).await
    }

    /// `GET /v1/sandboxes/{session_id}/processes/{process_id}/stream` -- live
    /// SSE stream of a background process's stdout, the streaming counterpart
    /// to [`get_process_output`](Client::get_process_output). Yields an
    /// [`ProcessStreamEvent::Output`] per new chunk, then a terminal
    /// [`ProcessStreamEvent::Exit`]. Bring `futures_util::StreamExt` into
    /// scope to consume it; the stream ends when the process finishes, the
    /// session is destroyed, or the caller drops it.
    pub fn stream_process_output(
        &self,
        session_id: &str,
        process_id: &str,
        since_offset: i64,
    ) -> Pin<Box<dyn Stream<Item = Result<ProcessStreamEvent, BoxkiteError>> + Send + 'static>> {
        let request_builder = self
            .request(
                Method::GET,
                &format!("/v1/sandboxes/{session_id}/processes/{process_id}/stream"),
            )
            .query(&[("since_offset", since_offset.to_string())]);

        Box::pin(async_stream::stream! {
            let mut event_source = match EventSource::new(request_builder) {
                Ok(event_source) => event_source,
                Err(err) => {
                    yield Err(BoxkiteError::Config(format!("failed to build stream request: {err}")));
                    return;
                }
            };

            while let Some(event) = event_source.next().await {
                match event {
                    Ok(Event::Open) => continue,
                    Ok(Event::Message(message)) => {
                        yield serde_json::from_str::<ProcessStreamEvent>(&message.data).map_err(BoxkiteError::from);
                    }
                    Err(reqwest_eventsource::Error::StreamEnded) => break,
                    Err(reqwest_eventsource::Error::InvalidStatusCode(status, response)) => {
                        let bytes = response.bytes().await.unwrap_or_default();
                        yield Err(crate::error::api_error_from_bytes(status.as_u16(), &bytes));
                        break;
                    }
                    Err(err) => {
                        yield Err(BoxkiteError::from(err));
                        break;
                    }
                }
            }
            event_source.close();
        })
    }

    /// `POST /v1/sandboxes/{session_id}/processes/{process_id}/input` --
    /// write to a tracked background process's stdin pipe (e.g. answering
    /// an interactive prompt).
    pub async fn send_process_input(
        &self,
        session_id: &str,
        process_id: &str,
        data: &str,
    ) -> Result<ProcessInputResult, BoxkiteError> {
        #[derive(Serialize)]
        struct Body<'a> {
            data: &'a str,
        }
        let builder = self
            .request(
                Method::POST,
                &format!("/v1/sandboxes/{session_id}/processes/{process_id}/input"),
            )
            .json(&Body { data });
        self.send(builder).await
    }

    /// `POST /v1/sandboxes/{session_id}/processes/{process_id}/stop` -- stop
    /// a tracked background process: SIGTERM, a short grace period, then
    /// SIGKILL if still alive.
    pub async fn stop_process(
        &self,
        session_id: &str,
        process_id: &str,
    ) -> Result<ProcessStopResult, BoxkiteError> {
        let builder = self.request(
            Method::POST,
            &format!("/v1/sandboxes/{session_id}/processes/{process_id}/stop"),
        );
        self.send(builder).await
    }
}
